File: core/agent_manager.py
"""
Agent Manager Module
Manages agent lifecycle and execution
"""

from typing import Dict, Any, Optional, List
from dataclasses import dataclass
import asyncio
import threading
import time
from core.logger import log
from core.agent_loader import AgentLoader


@dataclass
class AgentConfig:
    """Configuration for an agent"""
    name: str
    type: str
    enabled: bool = True
    max_concurrent: int = 1
    timeout: int = 300
    retry_count: int = 3


@dataclass
class AgentInstance:
    """Instance of a running agent"""
    id: str
    agent_class: Any
    config: AgentConfig
    status: str = "idle"
    last_activity: float = 0.0
    execution_count: int = 0


class AgentManager:
    """Manages agent lifecycle and execution"""

    def __init__(self):
        self.agents: Dict[str, AgentInstance] = {}
        self.agent_configs: Dict[str, AgentConfig] = {}
        self.loader = AgentLoader()
        self.running = False
        self._lock = threading.Lock()

    def load_agents(self, agents_dir: str = "agents") -> bool:
        """Load all agents from directory"""
        try:
            log.info(f"Loading agents from {agents_dir}")

            # Load agent modules
            modules = self.loader.load_all_agents(agents_dir)

            # Register agents from loaded modules
            for module_name, module in modules.items():
                if hasattr(module, 'Agent'):
                    agent_class = getattr(module, 'Agent')

                    # Create default config
                    config = AgentConfig(
                        name=getattr(agent_class, '__name__', module_name),
                        type=module_name
                    )

                    # Create agent instance
                    agent_instance = AgentInstance(
                        id=module_name,
                        agent_class=agent_class,
                        config=config
                    )

                    self.agents[module_name] = agent_instance
                    self.agent_configs[module_name] = config

            log.info(f"Successfully loaded {len(self.agents)} agents")
            return True

        except Exception as e:
            log.error(f"Failed to load agents: {e}")
            return False

    def get_agent(self, agent_id: str) -> Optional[AgentInstance]:
        """Get agent instance by ID"""
        return self.agents.get(agent_id)

    def get_all_agents(self) -> List[AgentInstance]:
        """Get all registered agents"""
        return list(self.agents.values())

    def start_agent(self, agent_id: str) -> bool:
        """Start an agent"""
        agent = self.get_agent(agent_id)
        if not agent:
            log.error(f"Agent not found: {agent_id}")
            return False

        with self._lock:
            if agent.status == "running":
                log.warning(f"Agent {agent_id} is already running")
                return False

            agent.status = "running"
            agent.last_activity = time.time()
            log.info(f"Started agent: {agent_id}")
            return True

    def stop_agent(self, agent_id: str) -> bool:
        """Stop an agent"""
        agent = self.get_agent(agent_id)
        if not agent:
            log.error(f"Agent not found: {agent_id}")
            return False

        with self._lock:
            if agent.status == "stopped":
                log.warning(f"Agent {agent_id} is already stopped")
                return False

            agent.status = "stopped"
            log.info(f"Stopped agent: {agent_id}")
            return True

    def execute_agent(self, agent_id: str, *args, **kwargs) -> Optional[Any]:
        """Execute an agent with given parameters"""
        agent = self.get_agent(agent_id)
        if not agent:
            log.error(f"Agent not found: {agent_id}")
            return None

        if agent.status != "running":
            log.error(f"Agent {agent_id} is not running")
            return None

        try:
            # Create agent instance
            agent_instance = agent.agent_class()

            # Execute the agent
            result = agent_instance.execute(*args, **kwargs)

            # Update execution count
            agent.execution_count += 1
            agent.last_activity = time.time()

            log.info(f"Agent {agent_id} executed successfully")
            return result

        except Exception as e:
            log.error(f"Agent {agent_id} execution failed: {e}")
            return None

    def get_agent_status(self, agent_id: str) -> Optional[Dict[str, Any]]:
        """Get status of an agent"""
        agent = self.get_agent(agent_id)
        if not agent:
            return None

        return {
            "id": agent.id,
            "status": agent.status,
            "execution_count": agent.execution_count,
            "last_activity": agent.last_activity,
            "config": agent.config.__dict__
        }

    def get_all_agent_status(self) -> Dict[str, Dict[str, Any]]:
        """Get status of all agents"""
        return {
            agent_id: self.get_agent_status(agent_id)
            for agent_id in self.agents.keys()
        }

    def reload_agent(self, agent_id: str) -> bool:
        """Reload an agent module"""
        agent = self.get_agent(agent_id)
        if not agent:
            log.error(f"Agent not found: {agent_id}")
            return False

        try:
            # Stop the agent first
            self.stop_agent(agent_id)

            # Reload the module
            module_name = agent.config.type
            success = self.loader.reload_agent(module_name)

            if success:
                log.info(f"Reloaded agent: {agent_id}")
                return True
            else:
                log.error(f"Failed to reload agent: {agent_id}")
                return False

        except Exception as e:
            log.error(f"Failed to reload agent {agent_id}: {e}")
            return False

    def start_all_agents(self) -> bool:
        """Start all agents"""
        try:
            for agent_id in self.agents.keys():
                self.start_agent(agent_id)
            log.info("Started all agents")
            return True
        except Exception as e:
            log.error(f"Failed to start all agents: {e}")
            return False

    def stop_all_agents(self) -> bool:
        """Stop all agents"""
        try:
            for agent_id in self.agents.keys():
                self.stop_agent(agent_id)
            log.info("Stopped all agents")
            return True
        except Exception as e:
            log.error(f"Failed to stop all agents: {e}")
            return False

    def cleanup(self):
        """Cleanup resources"""
        self.stop_all_agents()
        log.info("AgentManager cleanup completed")


# Example usage
if __name__ == "__main__":
    manager = AgentManager()
    success = manager.load_agents()
    if success:
        print(f"Loaded {len(manager.get_all_agents())} agents")
        print("AgentManager ready")
    else:
        print("Failed to load agents")

File: core/agent_manager.ts
import { EventEmitter } from 'events';
import { Agent, AgentConfig } from '../core/data_models';
import { AgentLoader } from '../core/agent_loader';

export interface AgentEvent {
  type: 'agent_registered' | 'agent_unregistered' | 'agent_status_changed' | 'agent_error';
  agentId: string;
  data?: any;
  timestamp: Date;
}

export class AgentManager extends EventEmitter {
  private agents: Map<string, Agent> = new Map();
  private loader: AgentLoader;
  private isMonitoring: boolean = false;
  private monitoringInterval: NodeJS.Timeout | null = null;

  constructor(pluginDir: string = 'plugins/agents') {
    super();
    this.loader = new AgentLoader(pluginDir);
    this.setupAgentMonitoring();
  }

  async initialize(): Promise<void> {
    try {
      // Load all existing agents
      const loadedAgents = await this.loader.loadAllAgents();

      for (const agent of loadedAgents) {
        this.registerAgent(agent);
      }

      // Set up file system watching for hot reload
      this.setupFileWatching();

      console.log(`Initialized AgentManager with ${loadedAgents.length} agents`);
    } catch (error) {
      console.error('Failed to initialize AgentManager:', error);
      throw error;
    }
  }

  private registerAgent(agent: Agent): void {
    this.agents.set(agent.id, agent);

    // Emit registration event
    this.emit('agent_registered', {
      type: 'agent_registered',
      agentId: agent.id,
      data: { agentName: agent.name, agentType: agent.type },
      timestamp: new Date()
    });

    console.log(`Registered agent: ${agent.name} (${agent.id})`);
  }

  private unregisterAgent(agentId: string): void {
    const agent = this.agents.get(agentId);
    if (agent) {
      this.agents.delete(agentId);

      // Emit unregistration event
      this.emit('agent_unregistered', {
        type: 'agent_unregistered',
        agentId: agentId,
        data: { agentName: agent.name, agentType: agent.type },
        timestamp: new Date()
      });

      console.log(`Unregistered agent: ${agent.name} (${agentId})`);
    }
  }

  private setupFileWatching(): void {
    try {
      const watcher = this.loader.watchAgentChanges(async (agentId, changeType) => {
        switch (changeType) {
          case 'added':
            console.log(`New agent file detected: ${agentId}`);
            const agents = await this.loader.loadAllAgents();
            for (const agent of agents) {
              if (!this.agents.has(agent.id)) {
                this.registerAgent(agent);
              }
            }
            break;
          case 'removed':
            if (this.agents.has(agentId)) {
              await this.unregisterAgent(agentId);
            }
            break;
          case 'modified':
            console.log(`Agent modified: ${agentId}`);
            // Hot reload the modified agent
            await this.reloadAgent(agentId);
            break;
        }
      });

      console.log('File watching enabled for agent hot reload');
    } catch (error) {
      console.error('Failed to setup file watching:', error);
    }
  }

  private setupAgentMonitoring(): void {
    this.monitoringInterval = setInterval(async () => {
      if (!this.isMonitoring) return;

      for (const [agentId, agent] of this.agents) {
        try {
          const status = await agent.getStatus();

          // Check for status changes
          if (status.lastStatus !== agent.lastStatus) {
            this.emit('agent_status_changed', {
              type: 'agent_status_changed',
              agentId: agentId,
              data: { oldStatus: agent.lastStatus, newStatus: status.lastStatus },
              timestamp: new Date()
            });

            agent.lastStatus = status.lastStatus;
          }
        } catch (error) {
          this.emit('agent_error', {
            type: 'agent_error',
            agentId: agentId,
            data: { error: error.message, stack: error.stack },
            timestamp: new Date()
          });
        }
      }
    }, 30000); // Check every 30 seconds
  }

  startMonitoring(): void {
    this.isMonitoring = true;
    console.log('Agent monitoring started');
  }

  stopMonitoring(): void {
    this.isMonitoring = false;
    console.log('Agent monitoring stopped');
  }

  async reloadAgent(agentId: string): Promise<boolean> {
    try {
      const agent = this.agents.get(agentId);
      if (!agent) {
        return false;
      }

      // Get agent path for reloading
      const agentPath = agent.path || '';
      const reloadedAgent = await this.loader.reloadAgent(agentId);

      if (reloadedAgent) {
        this.unregisterAgent(agentId);
        this.registerAgent(reloadedAgent);
        return true;
      }

      return false;
    } catch (error) {
      console.error(`Failed to reload agent ${agentId}:`, error);
      return false;
    }
  }

  async executeAgent(agentId: string, task: any): Promise<any> {
    try {
      const agent = this.agents.get(agentId);
      if (!agent) {
        throw new Error(`Agent ${agentId} not found`);
      }

      if (agent.status !== 'ready') {
        throw new Error(`Agent ${agentId} is not ready (status: ${agent.status})`);
      }

      // Update agent status to running
      agent.status = 'running';
      this.emit('agent_status_changed', {
        type: 'agent_status_changed',
        agentId: agentId,
        data: { oldStatus: 'ready', newStatus: 'running' },
        timestamp: new Date()
      });

      // Execute the task
      const result = await agent.execute(task);

      // Update agent status back to ready
      agent.status = 'ready';
      this.emit('agent_status_changed', {
        type: 'agent_status_changed',
        agentId: agentId,
        data: { oldStatus: 'running', newStatus: 'ready' },
        timestamp: new Date()
      });

      return result;
    } catch (error) {
      // Update agent status to error
      const agent = this.agents.get(agentId);
      if (agent) {
        agent.status = 'error';
        this.emit('agent_status_changed', {
          type: 'agent_status_changed',
          agentId: agentId,
          data: { oldStatus: 'running', newStatus: 'error' },
          timestamp: new Date()
        });
      }

      this.emit('agent_error', {
        type: 'agent_error',
        agentId: agentId,
        data: { error: error.message, task },
        timestamp: new Date()
      });

      throw error;
    }
  }

  getAgent(agentId: string): Agent | undefined {
    return this.agents.get(agentId);
  }

  getAllAgents(): Agent[] {
    return Array.from(this.agents.values());
  }

  getAgentsByType(agentType: string): Agent[] {
    return Array.from(this.agents.values()).filter(agent => agent.type === agentType);
  }

  getReadyAgents(): Agent[] {
    return Array.from(this.agents.values()).filter(agent => agent.status === 'ready');
  }

  async updateAgentConfig(agentId: string, config: Partial<AgentConfig>): Promise<boolean> {
    try {
      const existingConfig = await this.loader.loadAgentConfig(agentId);
      if (!existingConfig) {
        return false;
      }

      const updatedConfig = { ...existingConfig, ...config };
      return await this.loader.saveAgentConfig(agentId, updatedConfig);
    } catch (error) {
      console.error(`Failed to update agent config for ${agentId}:`, error);
      return false;
    }
  }

  async getAgentStats(): Promise<any> {
    const stats = {
      totalAgents: this.agents.size,
      byType: {} as Record<string, number>,
      byStatus: {} as Record<string, number>,
      averageUptime: 0,
      lastActivity: new Date()
    };

    for (const agent of this.agents.values()) {
      // Count by type
      stats.byType[agent.type] = (stats.byType[agent.type] || 0) + 1;

      // Count by status
      stats.byStatus[agent.status] = (stats.byStatus[agent.status] || 0) + 1;
    }

    return stats;
  }

  async shutdown(): Promise<void> {
    this.stopMonitoring();

    if (this.monitoringInterval) {
      clearInterval(this.monitoringInterval);
    }

    // Clean up all agents
    for (const [agentId, agent] of this.agents) {
      try {
        if (typeof agent.cleanup === 'function') {
          await agent.cleanup();
        }
        this.unregisterAgent(agentId);
      } catch (error) {
        console.error(`Error during agent ${agentId} cleanup:`, error);
      }
    }

    this.agents.clear();
    console.log('AgentManager shutdown complete');
  }
}

export default AgentManager;